queasars/job_shop_scheduling/scip_solver.py

```diff
@@ -0,0 +1,136 @@
+)
+
+
+class JSSPSCIPModelEncoder:
+    """
+    Encoding class used to encode a JobShopSchedulingProblemInstance as a pyscipopt SCIP solver model.
+    Needs queasars to be installed with the pyscipopt extra: pip install queasars[pyscipopt].
+    It also needs SCIP to be installed separately in the version 8.*.
+
+    :param jssp_instance: job shop scheduling problem instance to encode as a scip model
+    :type jssp_instance: JobShopSchedulingProblemInstance
+    """
+
+    def __init__(self, jssp_instance: JobShopSchedulingProblemInstance):
+        self._jssp_instance: JobShopSchedulingProblemInstance = jssp_instance
+        self._model_prepared: bool = False
+        self._model: Optional[Model] = None
+        self._optimization_var: Optional[Variable] = None
+        self._operation_start_variables: dict[Operation, Variable] = {}
+        self._machine_operations: dict[Machine, list[Operation]] = {}
+
+    def get_model(self) -> Model:
+        """
+        Creates and returns a pyscipopt SCIP optimization model. Subsequent calls return the cached model
+
+        :return: a pyscipopt SCIP optimization model
+        :rtype: Model
+        """
+        if not self._model_prepared:
+            self._model = Model()
+            self._prepare_variables()
+            self._prepare_constraints()
+            self._model_prepared = True
+        return self._model
+
+    def parse_solution(self, solution: Solution) -> JobShopSchedulingResult:
+        """
+        Parses a pyscipopt solution for this encoder's optimization model
+
+        :param solution: solution obtained by the SCIP solver
+        :type solution: Solution
+        :return: a JobShopSchedulingResult parsed from the solution
+        :rtype: JobShopSchedulingResult
+        """
+        job_schedules: dict[Job, tuple[ScheduledOperation, ...]] = {}
+        for job in self._jssp_instance.jobs:
+            scheduled_operations: list[ScheduledOperation] = []
+            for operation in job.operations:
+                start_time: int = int(solution[self._operation_start_variables[operation]])
+                scheduled_operations.append(ScheduledOperation(operation=operation, start=start_time))
+            job_schedules[job] = tuple(scheduled_operations)
+
+        return JobShopSchedulingResult(problem_instance=self._jssp_instance, schedule=job_schedules)
+
+    def _prepare_variables(self):
+        """
+        Creates the integer variables for the starting times of each operation and constrains them to be at least 0.
+        Also creates a integer variable to hold the makespan, which is constrained to be larger than the end times
+        of the last operations. Sets the optimization goal to minimize the makespan variable
+        """
+        # create auxiliary variable to hold the makespan as an optimization goal
+        self._optimization_var = self._model.addVar("optimization_var", vtype="INTEGER")
+        self._model.setObjective(self._optimization_var)
```

Comment:
Maybe check beforehand that `self._model` is not `None` anymore? This may happen if directly after initialization somebody calls this method here.

Reply:
I changed it so that the model is initialized in the constructor. This way it can never be None.

Reply:
Yes, that looks good, thank you!